dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java

```diff
@@ -25,272 +27,307 @@
+        this.program = new Program(new Memory(), format);
+    }
 
-    private final SourceLanguage format;
+    public static ProgramBuilder forArch(SourceLanguage format, Arch arch) {
+        final ProgramBuilder programBuilder = forLanguage(format);
+        programBuilder.program.setArch(arch);
+        return programBuilder;
+    }
 
-    public ProgramBuilder(SourceLanguage format) {
-        this.format = format;
+    public static ProgramBuilder forLanguage(SourceLanguage format) {
+        return new ProgramBuilder(format);
     }
-    
+
     public Program build() {
-        Program program = new Program(memory, format);
-        for (Function func : functions.values()) {
-            if (func instanceof Thread thread) {
-                addChild(thread.getId(), getOrCreateLabel(thread.getId(), "END_OF_T" + thread.getId()));
-                program.addThread(thread);
-            } else {
-                program.addFunction(func);
-            }
-            validateLabels(func);
+        for (Thread thread : program.getThreads()) {
+            final Label endOfThread = getEndOfThreadLabel(thread.getId());
+            // The terminator should not get inserted somewhere beforehand.
+            Verify.verify(endOfThread.getFunction() == null);
+            addChild(thread.getId(), endOfThread);
         }
+        id2FunctionsMap.values().forEach(this::validateFunction);
 
-        constants.forEach(program::addConstant);
-        program.setSpecification(ass);
-        program.setFilterSpecification(assFilter);
         EventIdReassignment.newInstance().run(program);
         program.getEvents().forEach(e -> e.setMetadata(new OriginalId(e.getGlobalId())));
         return program;
     }
 
-    public void initThread(String name, int tid) {
-        if(!functions.containsKey(tid)){
-            Skip threadEntry = EventFactory.newSkip();
-            functions.putIfAbsent(tid, new Thread(name, tid, threadEntry));
+    // ----------------------------------------------------------------------------------------------------------------
+    // Misc
+
+    public TypeFactory getTypeFactory() {
+        return types;
+    }
+
+    public ExpressionFactory getExpressionFactory() {
+        return expressions;
+    }
+
+    public void setAssert(AbstractAssert ass) {
+        program.setSpecification(ass);
+    }
+
+    public void setAssertFilter(AbstractAssert ass) {
+        program.setFilterSpecification(ass);
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Threads and Functions
+
+    public Thread newThread(String name, int tid) {
+        if(id2FunctionsMap.containsKey(tid)) {
+            throw new MalformedProgramException("Function or thread with id " + tid + " already exists.");
         }
+        final Thread thread = new Thread(name, DEFAULT_THREAD_TYPE, List.of(), tid, EventFactory.newSkip());
```

Comment:
Add a comment that we assume the main function does not have parameters. I am not sure we will ever get rid of this assumption, but better to have this properly documented.

Reply:
Actually, the parser does handle main with multiple input parameters, treating them all as unassigned registers (which are eventually treated as either 0 or non-deterministic, not sure).
I will add a comment that says that the thread type does not matter for now (I could set it to anything really).